### ISSUE TITLE
provide `this` context in SFCs when they are invoked via  `JSXMemberExpression` 

### ISSUE THREAD

**Issue by Artazor** [facebook/react#8714]
**[holy-war question alert!]**

Would it be possible to provide `this` context into SFCs when they are invoked via 
`JSXMemberExpression . JSXIdentifier` syntactic form?

My use-case involves unusual ES6 class usage. I'm using them as inheritable containers for SFC's, where individual components could be overridden in descendant classes.

Now I'm forced to autobind all view-methods:

```jsx
class Base {
    constructor() {
        // would like to avoid these:
        this.Short = this.Short.bind(this);
        this.Full = this.Full.bind(this);
    }
   Short() { return <span>{`${this}`}</span>; }
   Full() { return <div><this.Short /></div>; }
}

class Item extends Base {
    constructor(value) {
        super();
        this.value = value;
    }
    toString() {
        return this.value;
    }
}

var obj = new Item("A");
ReactDOM.render(<obj.Full />, document.getElementById("container"));
```

A bit more complete example could be found at: https://jsfiddle.net/69z2wepo/67046/

I understand that it is a Holy War question about pure functions and `this` in JavaScript.
However I'm thinking that this usage of `this` is legit. Just like a `children` in props.

It is way too ugly writing
```jsx
class Base {
   Short({self}) { return <span>{`${self}`}</span>; }
   Full({self}) { return <div><self.Short self={self} /></div>; }
}

class Item extends Base {
    constructor(value) {
        super();
        this.value = value;
    }
    toString() {
        return this.value;
    }
}

var obj = new Item("A");
ReactDOM.render(<obj.Full self={obj} />, document.getElementById("container"));
```
Isn't it?

**Comment by Artazor** [comment]
Updated example: https://jsfiddle.net/69z2wepo/67046/

**Comment by Artazor** [comment]
It would be also possible not to use `this`.
Just like `children` - provide some special member in props...
However it would be problematic - as it could clash with existing user code

**Comment by gaearon** [comment]
Thanks for taking the time to create an example.

I’m afraid this is indeed very unusual usage and not something we would encourage or add sugar to support. I would suggest you to stick to more traditional patterns of using React. Abstractions like this often end up hard to follow and unnecessarily complicated. Especially [when they involve inheritance](https://facebook.github.io/react/docs/composition-vs-inheritance.html).

For the reference, the “React way” of implementing this would look closer to:

```js
function List(props) {
  const { depth, items } = props;
  if (depth === 0) {
    return <span><b>[{items.length} items]</b></span>;
  }
  return (
    <ul>
      {items.map((item, index) =>
        <li key={index}>
          <Node depth={depth - 1} item={item} />
        </li>
      )}
    </ul>
  );
}

function Item(props) {
  const { depth, value } = props;
  if (depth === 0) {
    return <div><span>{value}</span></div>;
  }
  return <div>{value}</div>;
}

function Node(props) {
  const { item, depth } = props;
  if (Array.isArray(item)) {
    return <List items={item} depth={depth} />;
  }
  return <Item value={item} depth={depth} />;
}

var root = [
  'A',
  'B',
  [
    'C1',
    'C2',
    'C3',
    []
  ]
];

ReactDOM.render(<Node item={root} depth={2}/>, document.getElementById("container"));
```

Are you can see it reveals some unnecessary DOM wrappers which weren’t obvious in your example. (Although maybe this was intentional—it’s not clear.)

If you want to, you can further abstract *that* by using objects for your data and calling `toString()` on them. However I would encourage you not to mix inheritance-like techniques with React components.

I hope this helps somewhat!
